### ISSUE TITLE
Modifying java classes

### ISSUE THREAD

**Issue by micycle1** [INRIA/spoon#4664]
_This must have been addressed before but I can't find anything about it._

**Describe the bug**

When loading a class from its string and updating it (such as modifying method body), instances (`.newInstance()`) of the modified class reflect the changes. However, when loading the same class directly as a class source type, instances do not have the changes and nor is anything written to the `spooned-classes` folder.

**To Reproduce**

#### Loading from class directly -- not working!
```
Class<T> sketchClass = SimpleSketch.class; // (where <T extends PApplet>)

var spoon = new Launcher();
var f = spoon.getFactory();
var cls = f.Class().get(sketchClass);
cls.getMethodsByName("setup").get(0).getBody().addStatement(f.Code().createCodeSnippetStatement("print(\"test\")")));
System.out.println(cls.prettyprint()); // has the change
var instance = cls.newInstance();
instance.setup() // doesn't have the change!
```

#### Loading from a string of the same class -- working!
```
var spoon = new Launcher();
var f = spoon.getFactory();
Path path = Path.of("C:\\Users\\micyc\\eclipse-workspace-1\\pwatcher\\src\\main\\java\\pwatcher\\SimpleSketch.java");
String classStr = Files.readString(path);
cls = (CtClass<T>) Launcher.parseClass(classStr);
cls.getMethodsByName("setup").get(0).getBody().addStatement(f.Code().createCodeSnippetStatement("print(\"test\")")));
System.out.println(cls.prettyprint()); // has the change
var instance = cls.newInstance();
instance.setup() // still has the change
```

Input:

```
import processing.core.PApplet;

public class SimpleSketch extends PApplet {

	@Override
	public void settings() {
		size(800, 800, FX2D);
	}

	@Override
	public void setup() {
		background(255);
	}

	@Override
	public void draw() {
		background(0, 0, 255);
	}

}
```

In other words, how do you create a `CtClass` from a class type (rather than a string) that behaves properly?

**Comment by I-Al-Istannen** [comment]
That's not really possible as far as I can see. The only thing Spoon knows about your class in the first case is that it exists and can be queried via reflection. This is enough to build a model outline, but *not* enough to actually _synthesise a method body_. If you inspect your class, you will notice that all methods are empty. Spoon could have a look at the method bytecode, decompile it, and generate an AST that way - it sounds _really_ far out of the scope of this project though and is not really viable in the general case. 

As spoon does not have any source information nor a proper compilation unit and AST for your class, it has no code to compile and therefore does not produce any output class file. As no class file is present, spoon can not load it and instead falls back to the parent classloader, which returns the original class object you passed to spoon. That's why you see no changes.

In the second case, spoon actually has source information and a compilation unit and can make the changes you requested, compile it and load the newly compiled class file. Therefore, your code works.


> In other words, how do you create a CtClass from a class type (rather than a string) that behaves properly?

I believe the answer is: You cannot, as the problem isn't cleanly solvable *in general*, and especially not in the context of spoon (which analyses *source code*).

**Comment by monperrus** [comment]
This behavior is expected. CtClasses loaded from binary files (called shadow classes in Spoon) are meant to be read-only (introspection, not intercession).

In your snippet above, there should be an exception when you call `addStatement`, or even when you call `getBody()` (you cannot get the body of a shadow class).

Coming back to your question "how do you create a CtClass from a class type (rather than a string) that behaves properly?", the answer is by @slarse: look at module spoon-decompiler (not part of spoon-core)

**Comment by micycle1** [comment]
Thanks for the responses. Very clear.
